[cdap-585] spark to support filesets

anew opened the pull request:
Allows Spark to read and write ((time-)partitioned) file sets.
- adds support for InputFormatProvider and OutputFormatProvider in ExecutionSparkContext
- fixes a bug in SparkRuntimeService that was closing the datasets before committing them
- fixes a bug in SparkExecutionContext, which was throwing exception if dataset had no runtime args
- avoid double copying of the Hadoop conf in the SparkFacade
- adds tests for Java and Scala Spark programs that read and write each of the three kinds of file sets
- minor fix to RuntimeArguments.extractScope(), to always return a new map
- minor fix in FileSetArguments, to use FileInputFormat.INPUT_DIR instead of deprecated "mapred.input.dir"
 
The only thing that is not implemented is to call the onFailure() method of the DatasetOutputCommitter if the Spark job fails. For MapReduce it is obvious that this happens exactly for one dataset, at the end of the job. Spark can write multiple times, and it is not clear what the semantics are. If a write to one dataset succeeds, we call its onSuccess(), but if subsequently, a write to another dataset fails (or the processing in between), do we also have to call onFailure()? Or only for the one that failed? [CDAP-2983]

anew commented:
Build passed: http://builds.cask.co/browse/CDAP-DUT2208-4

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/common/RuntimeArguments.java:
```diff
@@ -89,11 +89,14 @@ private RuntimeArguments() {
    * @param name The name of the scope, e.g. "myTable"
    * @param arguments the runtime arguments of the enclosing scope
-   * @return a map that contains the arguments with and without prefix
+   * @return a new map that contains the arguments with and without prefix; or null if the input was null.
```
Will it return `null`? Should be marked with `@Nullable`?

chtyim replied:
Also, it may be easier to use if it returns empty map instead of null.

anew replied:
agreed

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java:
```diff
@@ -113,18 +118,95 @@ public ExecutionSparkContext(SparkSpecification specification, Id.Program progra
+
+    // first try if it is InputFormatProvider
+    Map<String, String> dsArgs = RuntimeArguments.extractScope(Scope.DATASET, datasetName, getRuntimeArguments());
+    Dataset dataset = getDataset(datasetName, dsArgs);
```
I think this dataset instance need to be closed in this method.

chtyim replied:
hum, maybe it can defer to when this context is closed.

chtyim replied:
But since Spark program can be iterative, meaning this method can be called many times. I think it's better to close it in this method.

anew replied:
However, then it will be closed again when the context is closed. Can we be sure that any dataset is ok with being closed repeatedly?

anew replied:
So, the getDatasets() remembers all datasets it instantiates in the datasets Map. If this is iterative, and the same dataset is used again, getDataset() will return the same instance. Which would be bad if I close it here. So, if I close it, I have to remove it from the datasets Map, for which currently there is no way provided. But if I do that, then rollbackTx() won't be called in case the program fails later. So I believe I have to keep it open.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java:
```diff
@@ -160,7 +242,8 @@ public ExecutionSparkContext(SparkSpecification specification, Id.Program progra
   @Override
   public synchronized <T extends Dataset> T getDataset(String name, Map<String, String> arguments) {
-    Map<String, String> datasetArgs = RuntimeArguments.extractScope(Scope.DATASET, name, getRuntimeArguments());
+    Map<String, String> datasetArgs = new HashMap<>();
```
If the contract of the `extractScope` always return a new HashMap, then we don't need to copy it one more time here.

anew replied:
ok

chtyim commented on cdap-app-fabric/src/test/scala/co/cask/cdap/internal/app/runtime/spark/ScalaFileCountProgram.scala:
```diff
@@ -0,0 +1,42 @@
+
+    // write to dataset
+    sc.writeToDataset(stringLengths, output, classOf[String], classOf[Integer])
+    sc.getOriginalSparkContext.asInstanceOf[org.apache.spark.SparkContext].stop()
```
No need to call stop from the spark program. The framework does it for you.

anew replied:
Oh, I did not pay attention. Just copied the existing ScalaCharCount program and modified it. I am fixing both now.

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/spark/SparkAppUsingFileSet.java:
```diff
@@ -0,0 +1,169 @@
+
+      // write the character count to dataset
+      context.writeToDataset(stringLengths, output, String.class, Integer.class);
+      ((JavaSparkContext) context.getOriginalSparkContext()).stop();
```
No need to call stop().

anew replied:
Same here. Will fix both.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java:
```diff
@@ -113,18 +118,95 @@ public ExecutionSparkContext(SparkSpecification specification, Id.Program progra
+
+    // first try if it is OutputFormatProvider
+    Map<String, String> dsArgs = RuntimeArguments.extractScope(Scope.DATASET, datasetName, getRuntimeArguments());
+    Dataset dataset = getDataset(datasetName, dsArgs);
```
Same as above, I think it's better to close this dataset instance in this method.

anew replied:
See my comment about this above.

chtyim replied:
Since dataset instances created in the `ExecutionSparkContext` are in long transaction, they won't get rollback, but instead the tx will just get invalided if the job fail.

chtyim replied:
By close, I mean `commitAndClose()` and remove from the local cache since the dataset instance created in here doesn't return back to user.

chtyim replied:
And to just get a new dataset instance without putting into the `datasets` instance cache, you can call `instantiateDataset` instead of `getDataset` (and that was why the `instantiateDataset` exists, which you can see both the `getBatchReadable` and `getBatchWritable` calls instead of `getDataset`).

anew commented:
Thanks for the review, @chtyim. I addressed all your comments. Please see especially my comments about closing the dataset. Thanks

anew commented:
@chtyim I made the suggested change. It is not that simple as for BatchWritable (using the delegate is easy if you only use a single method... here we need multiple). But it appears to work.

New build running here: http://builds.cask.co/browse/CDAP-DUT2208

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java:
```diff
@@ -113,18 +118,107 @@ public ExecutionSparkContext(SparkSpecification specification, Id.Program progra
+          // whatever went wrong, give the dataset a chance to handle the failure
+          if (dataset instanceof DatasetOutputCommitter) {
+            ((DatasetOutputCommitter) dataset).onFailure();
+          }
```
Do you need to throw? Otherwise the `onSuccess()` will also get executed.

anew replied:
absolutely. It must have been late...

chtyim commented:
Just one comment left. Rest LGTM.

anew commented:
@chtyim I fixed your last comment, also added a test case for a custom dataset that implements onSuccess/onFailure.

Also a fix for the AppFabricTestHelper: it failed when deploying an app with a custom dataset. When deploying the dataset module as part of the pipeline, it failed because the namespace data dir does not exist. Apparently the AppFabricTestHelper implicitly creates the default namespace (or its data dir) somewhere later in the pipeline I could not figure out where, but the fix is to create the namespace before deploying the app.

Please do one more review. Thanks

chtyim commented:
LGTM

anew commented:
@chtyim one last commit... please review -thanks

chtyim commented:
LGTM.. again.

anew commented:
Merged this before the build finished. But the last commit was only changing one test case, and that succeeds on my laptop. Merged this to unblock @rohitsinha54 for creating today's demo.
